### ISSUE TITLE
mixing ghosted and parallel vectors

### ISSUE THREAD

**Issue by manavbhatia** [libMesh/libmesh#2425]
I am hitting segmentation faults in `PetscVector::add()` [here](https://github.com/libMesh/libmesh/blob/f2e6647dc0216f99fa6f42c2d4ff9ca05d3ca6eb/src/numerics/petsc_vector.C#L374). This happens when `this->_type == GHOSTED` and `v_in._type == PARALLEL`. In this case [this](https://github.com/libMesh/libmesh/blob/f2e6647dc0216f99fa6f42c2d4ff9ca05d3ca6eb/src/numerics/petsc_vector.C#L371) returns `v_loc_vec` as a `nullptr` resulting in a segfault in call to [VecAXPY](https://github.com/libMesh/libmesh/blob/f2e6647dc0216f99fa6f42c2d4ff9ca05d3ca6eb/src/numerics/petsc_vector.C#L374). 

While other sections of this class and NumericVector make accommodations for mixing parallel and ghosted vectors, this method is not doing so. It would be nice to either enforce the similarity of vectors here or provide the necessary implementation to allow mixing of vectors.

**Comment by jwpeterson** [comment]
I agree we should probably check for matching types and throw an error when they don't. Would you mind making a PR for this since you seem to have a test case available? Or, if you want to try and figure out if it's possible to mix types (I'm not sure what this would entail) that would also be great!
